src/Juvix/Compiler/Core/Extra.hs

```diff
@@ -98,6 +102,44 @@ captureFreeVars fv
             | Just v <- s ^. at (u - k) -> NVar (Var i (v + k))
           m -> m
 
+-- captures all free variables of a node. It also returns the list of captured
+-- variables in left-to-right order: if snd is of the form λxλy... then fst is
+-- [x, y]
+captureFreeVarsCtx :: BinderList Binder -> Node -> ([(Var, Binder)], Node)
+captureFreeVarsCtx bl n =
+  let assocs = freeVarsCtx bl n
+   in (assocs, captureFreeVars (map (first (^. varIndex)) assocs) n)
+
+freeVarsCtx' :: BinderList Binder -> Node -> [Var]
+freeVarsCtx' bl = map fst . freeVarsCtx bl
+
+-- | the output list does not contain repeated elements and is sorted by variable index.
+-- TODO it can probably be fore efficient
```

Comment:
It can be more efficient and simpler: you just go through the BinderList and filter out the ones not present in the free variables of the body and previous binders?

Something like this should work, I think (treating BinderList as a list, I hope the shifting is correct but might've bungled it):

freeVarsCtx bl n = go 0 bl (freeVars n) []

go idx [] fvs acc = reverse acc
go idx (b : bs) fvs acc | Set.member idx fvs =
  let b' = shiftVar (idx + 1) b in
  go (idx + 1) bs (Set.union fvs (freeVars (b' ^. binderType))) ((idx, b') : acc)
go idx (_ : bs) fvs acc = go (idx + 1) bs fvs acc

Reply:
the error should be fixed. I have also simplified it a bit